### PR TITLE
Add the ability to run multiple elections

### LINKED ISSUE #579: Implement ability to run more than one election

### PR DESCRIPTION
**Pull request by leomassazza**
Fixes #579, #609 

This PR fixes the problem we had when we faced the Solidity limitation where it's not possible to add struct with nested mappings to arrays. To overcome this issue I changed the array to a mapping and added an uint index to identify the latest. As a side effect, since the uint index is used incrementally we can access (if needed) previous election data.

⚠️ This PR merges into the feature branch feature-ElectionModule and not to main. After all the small pieces of ElectionModule are done we need to merge the feature branch into main ⚠️

### CODE REVIEW

File: packages/core-modules/contracts/modules/CoreElectionModule.sol [hunk 1]
```diff
@@ -292,14 +295,20 @@ contract CoreElectionModule is IElectionModule, ElectionStorage, OwnableMixin {
     }
 
     function _findLessVoted(uint[] storage votes) private view returns (uint, uint) {
+        if (votes.length == 0) {
```
Comment (leomassazza) [inline packages/core-modules/contracts/modules/CoreElectionModule.sol:298]: Fix issue discovered while enabling the multi-staged tests

File: packages/core-modules/contracts/modules/CoreElectionModule.sol [hunk 2]
```diff
@@ -368,4 +381,8 @@ contract CoreElectionModule is IElectionModule, ElectionStorage, OwnableMixin {
     function getMaxProcessingBatchSize() external view override returns (uint256) {
         return _electionStore().maxProcessingBatchSize;
     }
+
+    function _electionData() private view returns (ElectionData storage) {
```
Comment (leomassazza) [inline packages/core-modules/contracts/modules/CoreElectionModule.sol:385]: Helper to get the ElectionData struct in the mapping.

File: packages/core-modules/contracts/storage/ElectionStorage.sol [hunk 3]
```diff
@@ -68,7 +68,11 @@ contract ElectionStorage {
         /**
          * @dev Voting data for the current election
          */
-        ElectionData electionData;
+        mapping(uint => ElectionData) electionData;
```
Comment (leomassazza) [inline packages/core-modules/contracts/storage/ElectionStorage.sol:71]: This is the core of this PR

File: packages/core-modules/test/contracts/modules/CoreElectionModule.Plurality.test.js [hunk 4]
```diff
@@ -69,6 +69,14 @@ describe('CoreElectionModule Count Votes using Simple Plurality strategy', () =>
         }
       });
 
+      after('cleanup voters', async () => {
```
Comment (leomassazza) [inline packages/core-modules/test/contracts/modules/CoreElectionModule.Plurality.test.js:72]: added to cleanup voters balance

File: packages/core-modules/test/contracts/modules/CoreElectionModule.Plurality.test.js [hunk 5]
```diff
@@ -152,7 +160,7 @@ describe('CoreElectionModule Count Votes using Simple Plurality strategy', () =>
     });
 
     // Skipped until we figure out how to run more than one election :/
-    describe.skip('when counting a large number of votes and candidates', async () => {
+    describe('when counting a large number of votes and candidates', async () => {
```
Comment (leomassazza) [inline packages/core-modules/test/contracts/modules/CoreElectionModule.Plurality.test.js:163]: it's alive!!!

File: packages/core-modules/test/contracts/modules/CoreElectionModule.Plurality.test.js [hunk 6]
```diff
@@ -242,10 +260,10 @@ describe('CoreElectionModule Count Votes using Simple Plurality strategy', () =>
               equal(nextEpochRepresentatives.length, 3);
               deepEqual(nextEpochRepresentatives, [
                 candidates[0].address,
-                candidates[98].address,
-                candidates[99].address,
+                candidates[9].address,
+                candidates[8].address,
               ]);
-              deepEqual(nextEpochRepresentativeVotesParsed, ['500', '130', '140']);
+              deepEqual(nextEpochRepresentativeVotesParsed, ['1350', '190', '180']);
```
Comment (leomassazza) [inline packages/core-modules/test/contracts/modules/CoreElectionModule.Plurality.test.js:266]: fix numbers with the right candidates number and balance

File: packages/core-modules/contracts/modules/CoreElectionModule.sol [hunk 7]
```diff
@@ -292,14 +295,20 @@ contract CoreElectionModule is IElectionModule, ElectionStorage, OwnableMixin {
...
     function _findLessVoted(uint[] storage votes) private view returns (uint, uint) {
+        if (votes.length == 0) {
+            return (0, 0);
+        }
+
         uint minVotes = votes[0];
         uint minIdx = 0;
-        for (uint idx = 0; idx < votes.length; idx++) {
+
+        for (uint idx = 1; idx < votes.length; idx++) {
```
Comment (leomassazza) [inline packages/core-modules/contracts/modules/CoreElectionModule.sol:305]: we can start comparing at idx 1

Comment (leomassazza) [inline packages/core-modules/contracts/modules/CoreElectionModule.sol:305]: we can start with idx 1

File: packages/core-modules/test/contracts/modules/CoreElectionModule.Plurality.test.js [hunk 8]
```diff
@@ -69,6 +69,14 @@ describe('CoreElectionModule Count Votes using Simple Plurality strategy', () =>
         }
       });
 
+      after('cleanup voters', async () => {
+        for (let i = 0; i < voters.length; i++) {
+          await ElectionToken.connect(voters[i]).burn(
+            await ElectionToken.balanceOf(voters[i].address)
+          );
+        }
```
Comment (mjlescano) [inline packages/core-modules/test/contracts/modules/CoreElectionModule.Plurality.test.js:77]: np:
```suggestion
        for (const voter of voters) {
          await ElectionToken.connect(voter).burn(
            await ElectionToken.balanceOf(voter.address)
          );
        }
```

File: packages/core-modules/test/contracts/modules/CoreElectionModule.Plurality.test.js [hunk 9]
```diff
@@ -178,6 +183,14 @@ describe('CoreElectionModule Count Votes using Simple Plurality strategy', () =>
...
```
Comment (mjlescano) [inline packages/core-modules/test/contracts/modules/CoreElectionModule.Plurality.test.js:191]: np:
```suggestion
        for (const voter of voters) {
          await ElectionToken.connect(voter).burn(
            await ElectionToken.balanceOf(voter.address)
          );
        }
```

**Review by mjlescano — APPROVED** [review]

File: packages/core-modules/contracts/storage/ElectionStorage.sol [hunk 10]
```diff
@@ -68,7 +68,11 @@ contract ElectionStorage {
...
+        /**
+         * @dev Index pointing to the latest electionDta
+         */
+        uint electionDataIndex;
```
Comment (ajsantander) [inline packages/core-modules/contracts/storage/ElectionStorage.sol:75]: Could we express this notion of "latest" in this property's name?
I.e. `latestElectionIdx` or `currentElectionIdx`.

File: packages/core-modules/contracts/modules/mocks/ElectionStorageMock.sol [hunk 11]
```diff
@@ -43,21 +43,22 @@ contract ElectionStorageMock is ElectionStorage {
...
     function getVoterVoteVotePowerMock(address addr) public view returns (uint) {
-        ElectionData storage electionData = _electionStore().electionData;
+        ElectionData storage electionData = _electionData();
         VoteData storage voteData = electionData.votes[electionData.votesIndexes[addr]];
         return voteData.votePower;
     }
 
-    function getEpochStart() public view returns (uint) {
-        return _electionStore().epochStart;
+    function _electionData() private view returns (ElectionData storage) {
```
Comment (ajsantander) [inline packages/core-modules/contracts/modules/mocks/ElectionStorageMock.sol:61]: What do you think about naming this `_currentElectionData()`? Here and in the non-mock contract of course.

File: packages/core-modules/contracts/storage/ElectionStorage.sol [hunk 12]
```diff
@@ -68,7 +68,11 @@ contract ElectionStorage {
         /**
          * @dev Voting data for the current election
```
Comment (ajsantander) [inline packages/core-modules/contracts/storage/ElectionStorage.sol:69]: This documentation may need a bit of tuning:
"Historical voting data" or something more semantically aligned.

File: packages/core-modules/contracts/modules/CoreElectionModule.sol [hunk 13]
```diff
@@ -246,13 +246,16 @@ contract CoreElectionModule is IElectionModule, ElectionStorage, OwnableMixin {
...
-        ElectionData storage electionData = store.electionData;
+        ElectionData storage electionData = _electionData();
 
         uint maxBatchSize = store.maxProcessingBatchSize;
         uint previousBatchIdx = electionData.processedBatchIdx;
         uint lessVotedCandidateVotes;
         uint lessVotedCandidateIdx;
 
+        // set initial values
+        (lessVotedCandidateIdx, lessVotedCandidateVotes) = _findLessVoted(electionData.nextEpochRepresentativeVotes);
```
Comment (ajsantander) [inline packages/core-modules/contracts/modules/CoreElectionModule.sol:257]: I don't like the property name `nextEpochRepresentative*`. We're calling these 'members', so why use 'representatives'? Feels inconsistent.

**Review by ajsantander — APPROVED** [review]
Awesome! Left a couple of nits, otherwise good to go 🚀

File: packages/core-modules/contracts/storage/ElectionStorage.sol (hunk 10 above)
Comment (leomassazza) [reply packages/core-modules/contracts/storage/ElectionStorage.sol]: Fixed

File: packages/core-modules/contracts/modules/mocks/ElectionStorageMock.sol (hunk 11 above)
Comment (leomassazza) [reply packages/core-modules/contracts/modules/mocks/ElectionStorageMock.sol]: Fixed

File: packages/core-modules/contracts/storage/ElectionStorage.sol (hunk 12 above)
Comment (leomassazza) [reply packages/core-modules/contracts/storage/ElectionStorage.sol]: Fixed

File: packages/core-modules/contracts/modules/CoreElectionModule.sol (hunk 13 above)
Comment (leomassazza) [reply packages/core-modules/contracts/modules/CoreElectionModule.sol]: Fixed (closing also #609  )

**Pull request merged**